lld/ELF/Arch/NanoMips.cpp

```diff
@@ -0,0 +1,562 @@
+  case R_NANOMIPS_PC25_S1: {
+    checkVal(loc, val - 4, 26, rel, rel.sym, 1, true);
+    writePcRel32<ELFT::TargetEndianness>(loc, val - 4, 25);
+    break;
+  }
+  case R_NANOMIPS_PC21_S1:
+    checkVal(loc, val - 4, 22, rel, rel.sym, 1, true);
+    writePcRel32<ELFT::TargetEndianness>(loc, val - 4, 21);
+    break;
+  case R_NANOMIPS_PC14_S1:
+    checkVal(loc, val - 4, 15, rel, rel.sym, 1, true);
+    writePcRel32<ELFT::TargetEndianness>(loc, val - 4, 14);
+    break;
+  case R_NANOMIPS_PC11_S1:
+    checkVal(loc, val - 4, 12, rel, rel.sym, 1, true);
+    writePcRel32<ELFT::TargetEndianness>(loc, val - 4, 11);
+    break;
+  case R_NANOMIPS_LO12:
+    writeValue32Shifted<ELFT::TargetEndianness>(loc, val, 12, 0);
+    break;
+  case R_NANOMIPS_LO4_S2:
+    checkVal(loc, val & 0xfff, 6, rel, rel.sym, 2, false);
+    writeValue16(loc, val >> 2, 4, 0);
+    break;
+  case R_NANOMIPS_GPREL19_S2:
+    checkVal(loc, val, 21, rel, rel.sym, 2, false);
+    writeValue32Shifted<ELFT::TargetEndianness>(loc, val, 19, 2);
+    break;
+  case R_NANOMIPS_GPREL18:
+    checkUInt(loc, val, 18, rel);
+    writeValue32Shifted<ELFT::TargetEndianness>(loc, val, 18, 0);
+    break;
+  case R_NANOMIPS_GPREL17_S1:
+    checkVal(loc, val, 18, rel, rel.sym, 1, false);
+    writeValue32Shifted<ELFT::TargetEndianness>(loc, val, 17, 1);
+    break;
+  case R_NANOMIPS_GPREL_LO12:
+    writeValue32Shifted<ELFT::TargetEndianness>(loc, val, 12, 0);
+    break;
+  default:
+    llvm_unreachable("unknown relocation");
+    break;
+  }
+}
+
+template <class ELFT> TargetInfo *elf::getNanoMipsTargetInfo() {
+  static NanoMips<ELFT> t;
+  return &t;
+}
+
+template <class ELFT>
+void NanoMips<ELFT>::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
+  const unsigned bits = config->is64 ? 64 : 32;
+  uint64_t secAddr = sec.getOutputSection()->addr;
+  if (auto *s = dyn_cast<InputSection>(&sec))
+    secAddr += s->outSecOff;
+  uint64_t valFromBefore = 0;
+  bool prevRelocOnlyCalculating = false;
+  for (auto it = sec.relocs().begin(), end = sec.relocs().end(); it != end;
+       it++) {
+
+    const Relocation &rel = *it;
+    if (rel.expr == R_RELAX_HINT)
+      continue;
+
+    uint8_t *loc = buf + rel.offset;
+    uint64_t val = 0;
+    if (!prevRelocOnlyCalculating)
+      val = SignExtend64(sec.getRelocTargetVA(sec.file, rel.type, rel.addend,
+                                              secAddr + rel.offset, *rel.sym,
+                                              rel.expr),
+                         bits);
+
+    // If prev reloc was part of composite relocations
+    // use its val for this reloc's addend
+    // TODO: See if maybe addend should be
+    // added to valFromBefore
+    else
+      val = SignExtend64(sec.getRelocTargetVA(sec.file, rel.type, valFromBefore,
+                                              secAddr + rel.offset, *rel.sym,
+                                              rel.expr),
+                         bits);
+
+    auto next = it + 1;
```

Comment:
I think the code in general does not follow https://llvm.org/docs/CodingStandards.html

Reply:
The overall code in the whole file, or this instance auto next = it + 1?